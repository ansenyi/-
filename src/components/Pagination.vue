<template>
  <div class="pagination">
    <label for="page" class="select-label">
      <select id="page" class="select-size" ref="selectSize" @change="selectSize">
        <option value="5">每页10条</option>
        <option value="20">每页20条</option>
        <option value="30">每页30条</option>
      </select>
      <i id="first" class="select-down icon-down"></i>
    </label>
    <ul class="page-list">
      <li class="prev" v-show="curPage !== 1" @click="prevPage">
        <span><i class="icon-prev"></i></span>
      </li>
      <li class="first" v-show="totalPages > 1" @click="goPage(1)">
        <span :class="{'active': curPage === 1}">1</span>
      </li>
      <li class="ellipsis" v-show="prevClipped"><span>...</span></li>
      <li v-for="index in pages" :key="index" @click="goPage(index)">
        <span :class="{'active': curPage === index}">{{ index }}</span>
      </li>
      <li class="ellipsis" v-show="nextClipped"><span>...</span></li>
      <li class="last" @click="goPage(totalPages)">
        <span :class="{'active': curPage === totalPages}">{{ totalPages }}</span>
      </li>
      <li class="next" v-show="curPage !== totalPages" @click="nextPage">
        <span><i class="icon-next"></i></span>
      </li>
    </ul>
  </div>
</template>
<script>
  export default {
    name: 'pagination',
    props: {
      totalPages: {
        type: Number,
        default: 1,
        validator: i => i > 0
      },
      current: {
        type: Number,
        default: 1,
        validator: i => i > 0
      }
    },
    data () {
      return {
        nextClipped: false,
        prevClipped: false,
        curPage: this.current
      }
    },
    computed: {
      pages () {
        let ret = []
        this.curPage = this.current
        if (this.curPage > 3) {
          // 当前页码大于三时，显示当前页码的前2个
          ret.push(this.curPage - 2)
          ret.push(this.curPage - 1)
          this.prevClipped = this.curPage > 4
        } else {
          this.prevClipped = false
          for (let i = 2; i < this.curPage; i++) {
            ret.push(i)
          }
        }
        if (this.curPage !== this.totalPages && this.curPage !== 1) {
          ret.push(this.curPage)
        }
        if (this.curPage < (this.totalPages - 2)) {
          // 显示当前页码的后2个
          ret.push(this.curPage + 1)
          ret.push(this.curPage + 2)
          // 当前页与最后一页差距3以上时显示省略号
          this.nextClipped = this.curPage < (this.totalPages - 3)
        } else {
          this.nextClipped = false
          for (let i = (this.curPage + 1); i < this.totalPages; i++) {
            ret.push(i)
          }
        }
        return ret
      }
    },
    watch: {
      curPage (newVal) {
        console.log('当前页面currentPage：' + newVal)
        this.$emit('toPage', newVal)
        // this.$forceUpdate()
      }
    },
    methods: {
      selectSize () {
        let select = this.$refs.selectSize
        let index = select.selectedIndex
        let size = select[index].value
        this.$emit('changeSize', size)
      },
      prevPage () {
        this.curPage--
      },
      nextPage () {
        this.curPage++
      },
      goPage (index) {
        if (index !== this.curPage) {
          this.curPage = index
        }
      }
    }
  }
</script>

<style lang="stylus" scoped>
  @import '../common/stylus/mixin.styl'
  .pagination
    display flex
    justify-content center
    align-items center
    margin-top 20px
    .select-label
      select-label()
      &:before
        top 0
        right 0
        width 94px
        height 26px
        border-radius 3px
      .select-size
        height 26px
        color #444
        border 1px solid #999
        border-radius 3px
      .select-down
        top 4px
        right 4px
        color #999
    .page-list
      display flex
      li
        list-style none
        color #666
        //padding 0 3px
        &:not(.ellipsis)
          span
            display block
            min-width 24px
            height 25px
            margin 0 4px
            line-height 25px
            text-align center
            border 1px solid #999
            border-radius 3px
            text-shadow 0 1px 0 #fff
            cursor pointer
            background #fff
            i
              line-height 25px
          .active, span:hover
            color #fff
            background #0077d5
            /*border 1px solid #0077d5*/
</style>
