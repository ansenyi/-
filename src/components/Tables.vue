<template>
  <div class="tables">
    <div class="search">
      <i class="input-search icon-search" @click="searchLike"></i>
      <input type="text" v-model="searchText" ref="search" :placeholder="placeholder" @keyup.enter="searchLike" autocomplete="">
      <i class="input-cancel icon-cancel" v-show="showCancel" @click="searchText = '';searchLike()"></i>
    </div>
    <table>
      <caption class="count">
        <slot name="count"></slot>
      </caption>
      <thead @click="sortClick($event)">
	      <tr>
	        <th v-for="(thead, index) in theads" :key="index">
	          <span>{{ thead.title }}</span>
	          <a><i class="icon-asc" v-if="index !== theads.length-1 && thead.prop !== 'ip'" data-direction="asc" :data-prop="thead.prop"></i></a>
	          <a><i class="icon-desc" v-if="index !== theads.length-1 && thead.prop !== 'ip'" data-direction="desc" :data-prop="thead.prop"></i></a>
	        </th>
	      </tr>
      </thead>
      <tbody @click="removeId($event)">
	      <tr v-for="(items, index) in tableData" :class="{ 'grey': items[5] === '未连接', 'red': items[1] === 'Untitled'}" :key="index">
	        <td v-for="(item, index) in items" :class="{'occu-red': ($route.name === 'DevicePreview' && index === items.length - 1 && item.replace('%', '') > 100), 'occu-yellow': ($route.name === 'DevicePreview' && index === items.length - 1 && item.replace('%', '') > 80 && item.replace('%', '') < 100)}"
	            v-if="!($route.name === 'PersonPreview' && index === 0)" :key="index">{{ item }}
	        </td>
	        <td>
	          <span v-if="$route.name === 'RolePreview'" @click="eEditShow(items[0])"><i class="icon-edit"></i></span>
	          <router-link v-else :to="encodeURI(settingPath + '/' + items[0])"><i class="icon-edit"></i></router-link>
	          <span v-if="$route.name === 'DevicePreview' || $route.name === 'PersonPreview'" @click="detailShow(items[0])"><i class="icon-list"></i></span>
	          <span v-if="$route.name === 'DevicePreview' || $route.name === 'PersonPreview'" @click="tableShow(items[0], items[1])"><i class="icon-table"></i></span>
	          <span><i class="icon-remove" :data-id="items[0]" :data-name="items[1]"></i></span>
	        </td>
	      </tr>
      </tbody>
    </table>
    <pagination :totalPages="totalPages ? totalPages : 1" :current="curPage + 1" @toPage="toPage" @changeSize="changeSize"></pagination>
  </div>
</template>

<script>
  import Pagination from './Pagination.vue'
  export default {
    name: 'tables',
    props: {
      placeholder: {
        type: String,
        default: ''
      },
      theads: {
        type: Array,
        default: () => []
      },
      tableData: {
        type: Array,
        default: () => []
      },
      settingPath: {
        type: String,
        default: ''
      },
      totalPages: {
        type: Number,
        default: 1
      },
      curPage: {
        type: Number,
        default: 1
      }
    },
    data () {
      return {
        searchText: ''
      }
    },
    computed: {
      showCancel () {
        return this.searchText !== ''
      }
    },
    methods: {
      sortClick (e) {
        let target = e ? e.target : window.event.srcElement
        let prop = target.dataset.prop
        if (target.dataset.direction === 'asc') {
          console.log('按asc顺序排')
          this.$emit('updateData', undefined, undefined, 'asc', prop)
        } else if (target.dataset.direction === 'desc') {
          console.log('按desc顺序排')
          this.$emit('updateData', undefined, undefined, 'desc', prop)
        }
      },
      searchLike () {
        // this.$refs.search.blur()使搜索框失去焦点
        this.$emit('updateData', undefined, undefined, undefined, undefined, this.searchText)
      },
      removeId (e) {
        let target = e ? e.target : window.event.srcElement
        if (target.className === 'icon-remove' && target.dataset.id) {
          let id = target.dataset.id
          let name = target.dataset.name
          console.log('id=' + id)
          this.$emit('removeId', id, name)
        }
      },
      toPage (curPage) {
        this.$emit('updateData', curPage - 1)
      },
      changeSize (size) {
        this.$emit('updateData', undefined, size)
      },
      detailShow (id) {
        console.log('进行详情显示')
        if (this.$route.name === 'DevicePreview') {
          this.$emit('showDeviceDetail', id)
        } else if (this.$route.name === 'PersonPreview') {
          this.$emit('showPersonDetail', id)
        }
      },
      eEditShow (id) {
        console.log('显示edit编辑页面')
        if (this.$route.name === 'RolePreview') {
          this.$emit('showRoleEdit', id)
        }
      },
      tableShow (id, name) {
        if (this.$route.name === 'DevicePreview') {
          this.$emit('showDeviceTable', id, name)
        } else if (this.$route.name === 'PersonPreview') {
          this.$emit('showPersonTable', id, name)
        }
      }
    },
    components: {
      Pagination
    }
  }
</script>

<style lang="stylus" scoped>
  .occu-red
    color #d84c4c
  .occu-yellow
    color #0077d5
  .tables
    margin-top -50px
    .grey
      color #999 !important
    .red
      color #d84c4c
    .search
      position relative
      width 250px
      height 32px
      margin 40px auto 30px 10%
      font-size 16px
      line-height 32px
      .input-search
        position absolute
        top 0
        left 0
        font-size 30px
        line-height 32px
        cursor pointer
        &:active
          color #ccc
      .input-cancel
        position absolute
        top 0
        right 5px
        font-size 20px
        line-height 32px
        color #ccc
        cursor pointer
        &:hover
          color #888
      input
        width 100%
        padding 2px 30px
        box-sizing border-box
        line-height 24px
    table, th, td
      border 1.5px solid #eee
    table
      border-collapse collapse
      margin 0 auto
      max-width 95%
      .count
        padding-right 5px
        text-align right
        font-size 14px
        span
          padding-left 30px
      th, td
        max-width 300px
        height 38px
        /*color #888888*/
      th
        position relative
        min-width 95px
        font-weight normal
        color #000
        background #fafafa
        background-clip padding-box
        padding 0 10px
        a
          position absolute
          right 6px
          font-size 10px
          cursor pointer
          &:first-of-type
            top 6px
          &:last-of-type
            bottom 6px
          &:active
            color #ccc
      td
        text-align center
        padding 0 10px
        background #fff
        a
          text-decoration none
        span
          cursor pointer
        i
          color #2dc5fa
          font-size 28px
        &:not(:last-of-type)
          white-space pre-wrap
          word-break break-all
        &:last-of-type
          white-space nowrap
      .control-disabled
        cursor default
        i
          color #ccc
</style>
