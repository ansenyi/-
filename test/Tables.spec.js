/* eslint-disable no-undef,no-unused-vars */
import 'babel-polyfill'
import jest from 'jest-mock'
import { shallow, mount } from 'vue-test-utils'
import Tables from '@/components/Tables.vue'
import Pagination from '@/components/Pagination.vue'

describe('tables', () => {
  function initDeviceTable () {
    const $route = {
      name: 'DevicePreview'
    }
    const cmp = shallow(Tables, {
      mocks: {
        $route
      },
      propsData: {
        placeholder: '按设备名称搜索',
        theads: [{prop: 'deviceId', title: '设备编号'}, {prop: 'deviceName', title: '设备名称'}, {prop: 'protectLock', title: '防拆开关'}, {prop: 'hardwareVersion', title: '硬件版本号'}, {prop: 'online', title: '连接状态'}, {prop: 'ip', title: '通讯地址'}, {title: '设备管理'}],
        tableData: [[1, 'a', '开', 'v1', '未连接', '123'], [2, 'Untitled', '关', 'v1', '已连接', '456']],
        settingPath: '/device/setting'
      }
    })
    return cmp
  }
  function initPersonTable () {
    const $route = {
      name: 'PersonPreview'
    }
    const cmp = shallow(Tables, {
      mocks: {
        $route
      },
      propsData: {
        placeholder: '按姓名搜索',
        theads: [{prop: 'userName', title: '姓名'}, {prop: 'employeeId', title: '工号'}, {prop: 'carNO', title: '卡号'}, {prop: 'orgEntity[\'orgName\']', title: '部门'}, {prop: 'roleEntity[\'roleName\']', title: '角色'}, {title: '人员信息操作'}],
        tableData: [[1, 'a', 11, 111, '营销部', '未分配'], [2, 'b', 22, 222, '已连接', '456']],
        settingPath: '/person/setting'
      }
    })
    return cmp
  }
  it('props type', () => {
    const cmp = shallow(Tables)
    const placeholder = cmp.vm.$options.props.placeholder
    const theads = cmp.vm.$options.props.theads
    const tableData = cmp.vm.$options.props.tableData
    const settingPath = cmp.vm.$options.props.settingPath
    const totalPages = cmp.vm.$options.props.totalPages
    const curPage = cmp.vm.$options.props.curPage
    expect(placeholder.type).toBe(String)
    expect(theads.type).toBe(Array)
    expect(tableData.type).toBe(Array)
    expect(settingPath.type).toBe(String)
    expect(totalPages.type).toBe(Number)
    expect(curPage.type).toBe(Number)
  })
  it('set device table props', () => {
    const cmp = initDeviceTable()
    expect(cmp.element).toMatchSnapshot()
  })
  it('set person table props', () => {
    const cmp = initPersonTable()
    expect(cmp.element).toMatchSnapshot()
  })
  it('search input', () => {
    const cmp = initDeviceTable()
    cmp.setData({searchText: '159'})
    // cmp.update()
    expect(cmp.vm.searchText).toBe('159')
    expect(cmp.contains('.input-cancel')).toBeTruthy()
    cmp.find('.input-cancel').trigger('click')
    // cmp.update()
    expect(cmp.vm.searchText).toBe('')
  })
  describe('updateData', () => {
    const cmp = initDeviceTable()
    const stub = jest.fn()
    cmp.vm.$on('updateData', stub)
    it('searchLike', () => {
      cmp.find('input[type=text]').trigger('blur')
      cmp.update()
      expect(stub).toBeCalled()
    })
    it('sortClick', () => {
      cmp.find('th a i').trigger('click')
      cmp.update()
      expect(stub).toBeCalled()
    })
    it('remove', () => {
      const stup = jest.fn()
      cmp.vm.$on('removeId', stup)
      cmp.find('.icon-remove').trigger('click')
      cmp.vm.$nextTick(() => {
        expect(stup).toBeCalledWith('1')
      })
      cmp.find('tbody tr').trigger('click')
      cmp.vm.$nextTick(() => {
        expect(stup).not.toBeCalled()
      })
      cmp.find('tbody tr td').trigger('click')
      cmp.vm.$nextTick(() => {
        expect(stup).not.toBeCalled()
      })
      cmp.find('tbody tr td router-link').trigger('click')
      cmp.vm.$nextTick(() => {
        expect(stup).not.toBeCalled()
      })
      cmp.find('tbody tr td span').trigger('click')
      cmp.vm.$nextTick(() => {
        expect(stup).not.toBeCalled()
      })
    })
    describe('pagination component', () => {
      const wrapper = mount(Tables, {
        propsData: {
          totalPages: 5,
          curPage: 1
        }
      })
      const pagination = wrapper.find(Pagination)
      it('next page', () => {
        pagination.find('.next').trigger('click')
        cmp.vm.$nextTick(() => {
          expect(stub).toBeCalledWith('1')
        })
      })
      it('to page 3', () => {
        pagination.find('li:nth-of-type(5)').trigger('click')
        cmp.vm.$nextTick(() => {
          expect(stub).toBeCalledWith('1')
        })
      })
    })
  })
})
