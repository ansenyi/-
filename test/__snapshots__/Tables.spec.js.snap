// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tables set device table props 1`] = `
<div
  class="tables"
>
  <section
    class="search"
  >
    <i
      class="input-search icon-search"
    />
     
    <input
      autocomplete="off"
      placeholder="按设备名称搜索"
      type="text"
    />
     
    <i
      class="input-cancel icon-cancel"
      style="display: none;"
    />
  </section>
   
  <table>
    <caption
      class="count"
    />
     
    <thead>
      <tr>
        <th>
          <span>
            设备编号
          </span>
           
          <a>
            <i
              class="icon-asc"
              data-direction="asc"
              data-prop="deviceId"
            />
          </a>
           
          <a>
            <i
              class="icon-desc"
              data-direction="desc"
              data-prop="deviceId"
            />
          </a>
        </th>
        <th>
          <span>
            设备名称
          </span>
           
          <a>
            <i
              class="icon-asc"
              data-direction="asc"
              data-prop="deviceName"
            />
          </a>
           
          <a>
            <i
              class="icon-desc"
              data-direction="desc"
              data-prop="deviceName"
            />
          </a>
        </th>
        <th>
          <span>
            防拆开关
          </span>
           
          <a>
            <i
              class="icon-asc"
              data-direction="asc"
              data-prop="protectLock"
            />
          </a>
           
          <a>
            <i
              class="icon-desc"
              data-direction="desc"
              data-prop="protectLock"
            />
          </a>
        </th>
        <th>
          <span>
            硬件版本号
          </span>
           
          <a>
            <i
              class="icon-asc"
              data-direction="asc"
              data-prop="hardwareVersion"
            />
          </a>
           
          <a>
            <i
              class="icon-desc"
              data-direction="desc"
              data-prop="hardwareVersion"
            />
          </a>
        </th>
        <th>
          <span>
            连接状态
          </span>
           
          <a>
            <!---->
          </a>
           
          <a>
            <!---->
          </a>
        </th>
        <th>
          <span>
            通讯地址
          </span>
           
          <a>
            <!---->
          </a>
           
          <a>
            <!---->
          </a>
        </th>
        <th>
          <span>
            设备管理
          </span>
           
          <a>
            <!---->
          </a>
           
          <a>
            <!---->
          </a>
        </th>
      </tr>
    </thead>
     
    <tbody>
      <tr
        class="grey"
      >
        <td>
          1
        </td>
        <td>
          a
        </td>
        <td>
          开
        </td>
        <td>
          v1
        </td>
        <td>
          未连接
        </td>
        <td>
          123
        </td>
         
        <td
          class="control-disabled"
        >
          <span>
            <i
              class="icon-edit"
            />
          </span>
           
          <span>
            <i
              class="icon-remove"
            />
          </span>
        </td>
      </tr>
      <tr
        class="red"
      >
        <td>
          2
        </td>
        <td>
          Untitled
        </td>
        <td>
          关
        </td>
        <td>
          v1
        </td>
        <td>
          已连接
        </td>
        <td>
          456
        </td>
         
        <td>
          <router-link
            to="/device/setting/2"
          >
            <i
              class="icon-edit"
            />
          </router-link>
           
          <span>
            <i
              class="icon-remove"
              data-id="2"
            />
          </span>
        </td>
      </tr>
    </tbody>
  </table>
   
  <!---->
</div>
`;

exports[`tables set person table props 1`] = `
<div
  class="tables"
>
  <section
    class="search"
  >
    <i
      class="input-search icon-search"
    />
     
    <input
      autocomplete="off"
      placeholder="按姓名搜索"
      type="text"
    />
     
    <i
      class="input-cancel icon-cancel"
      style="display: none;"
    />
  </section>
   
  <table>
    <caption
      class="count"
    />
     
    <thead>
      <tr>
        <th>
          <span>
            姓名
          </span>
           
          <a>
            <i
              class="icon-asc"
              data-direction="asc"
              data-prop="userName"
            />
          </a>
           
          <a>
            <i
              class="icon-desc"
              data-direction="desc"
              data-prop="userName"
            />
          </a>
        </th>
        <th>
          <span>
            工号
          </span>
           
          <a>
            <i
              class="icon-asc"
              data-direction="asc"
              data-prop="employeeId"
            />
          </a>
           
          <a>
            <i
              class="icon-desc"
              data-direction="desc"
              data-prop="employeeId"
            />
          </a>
        </th>
        <th>
          <span>
            卡号
          </span>
           
          <a>
            <i
              class="icon-asc"
              data-direction="asc"
              data-prop="carNO"
            />
          </a>
           
          <a>
            <i
              class="icon-desc"
              data-direction="desc"
              data-prop="carNO"
            />
          </a>
        </th>
        <th>
          <span>
            部门
          </span>
           
          <a>
            <i
              class="icon-asc"
              data-direction="asc"
              data-prop="orgEntity['orgName']"
            />
          </a>
           
          <a>
            <i
              class="icon-desc"
              data-direction="desc"
              data-prop="orgEntity['orgName']"
            />
          </a>
        </th>
        <th>
          <span>
            角色
          </span>
           
          <a>
            <i
              class="icon-asc"
              data-direction="asc"
              data-prop="roleEntity['roleName']"
            />
          </a>
           
          <a>
            <i
              class="icon-desc"
              data-direction="desc"
              data-prop="roleEntity['roleName']"
            />
          </a>
        </th>
        <th>
          <span>
            人员信息操作
          </span>
           
          <a>
            <!---->
          </a>
           
          <a>
            <!---->
          </a>
        </th>
      </tr>
    </thead>
     
    <tbody>
      <tr
        class=""
      >
        <!---->
        <td>
          a
        </td>
        <td>
          11
        </td>
        <td>
          111
        </td>
        <td>
          营销部
        </td>
        <td>
          未分配
        </td>
         
        <td>
          <router-link
            to="/person/setting/1"
          >
            <i
              class="icon-edit"
            />
          </router-link>
           
          <span>
            <i
              class="icon-remove"
              data-id="1"
            />
          </span>
        </td>
      </tr>
      <tr
        class=""
      >
        <!---->
        <td>
          b
        </td>
        <td>
          22
        </td>
        <td>
          222
        </td>
        <td>
          已连接
        </td>
        <td>
          456
        </td>
         
        <td>
          <router-link
            to="/person/setting/2"
          >
            <i
              class="icon-edit"
            />
          </router-link>
           
          <span>
            <i
              class="icon-remove"
              data-id="2"
            />
          </span>
        </td>
      </tr>
    </tbody>
  </table>
   
  <!---->
</div>
`;
